oak_attestation_verification/src/verifier.rs

```diff
@@ -120,7 +129,53 @@ fn verify_dice_chain(evidence: &Evidence) -> anyhow::Result<()> {
         verifying_key = cose_key_to_verifying_key(&cose_key).map_err(|msg| anyhow::anyhow!(msg))?;
     }
 
-    Ok(())
+    // Verify application keys in evidence and extract public keys. The verifying key comes from the
+    // last DICE layer.
+    let appl_keys = evidence
+        .application_keys
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no application fields in evidence"))?;
+
+    // Process encryption certificate.
+    let encryption_cert =
+        coset::CoseSign1::from_slice(&appl_keys.encryption_public_key_certificate)
+            .map_err(|_cose_err| anyhow::anyhow!("could not parse encryption certificate"))?;
+    encryption_cert.verify_signature(ADDITIONAL_DATA, |signature, contents| {
+        let sig = Signature::from_slice(signature)?;
+        verifying_key.verify(contents, &sig)
+    })?;
+    let encryption_payload = encryption_cert
+        .payload
+        .ok_or_else(|| anyhow::anyhow!("no encryption cert payload"))?;
+    let encryption_claims = ClaimsSet::from_slice(&encryption_payload)
+        .map_err(|_cose_err| anyhow::anyhow!("could not parse encryption claims set"))?;
+    let encryption_cose_key =
+        get_public_key_from_claims_set(&encryption_claims).map_err(|msg| anyhow::anyhow!(msg))?;
+    let encryption_public_key =
+        cose_key_to_hpke_public_key(&encryption_cose_key).map_err(|msg| anyhow::anyhow!(msg))?;
+
+    // Process signing certificate.
+    let signing_cert = coset::CoseSign1::from_slice(&appl_keys.signing_public_key_certificate)
+        .map_err(|_cose_err| anyhow::anyhow!("could not parse encryption certificate"))?;
+    signing_cert.verify_signature(ADDITIONAL_DATA, |signature, contents| {
+        let sig = Signature::from_slice(signature)?;
+        verifying_key.verify(contents, &sig)
+    })?;
+    let signing_payload = signing_cert
+        .payload
+        .ok_or_else(|| anyhow::anyhow!("no signing cert payload"))?;
+    let signing_claims = ClaimsSet::from_slice(&signing_payload)
+        .map_err(|_cose_err| anyhow::anyhow!("could not parse signing claims set"))?;
+    let signing_cose_key: CoseKey =
+        get_public_key_from_claims_set(&signing_claims).map_err(|msg| anyhow::anyhow!(msg))?;
+    let signing_verifying_key =
+        cose_key_to_verifying_key(&signing_cose_key).map_err(|msg| anyhow::anyhow!(msg))?;
+    let signing_public_key = signing_verifying_key.to_public_key_der()?.into_vec();
```

Comment:
I think this should use `to_sec1_bytes`, rather than `to_public_key_der` since we don't generally use DER encoding for public key serialisation.

Reply:
Done.